### PR TITLE
Enable configuration of auto instrumentation securityContext for init containers

### PR DESCRIPTION
**Pull request by mschier**
### What does this PR do?

Add an option to configure a [Security Context](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#securitycontext-v1-core) to be applied to all `initContainers` that the auto instrumentation feature creates to inject libraries.

### Motivation

In our organization we use a controlled Kubernetes cluster where we have enforced a `Restricted` [Pod Security Standard](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted), which requires all containers to at least specify certain keys in their [Security Context](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#securitycontext-v1-core). This de-facto prevents us from utilizing the auto instrumentation feature, as the `initContainers` added specifies no `securityContext` at all.

This change would allow for generic configuration, also ensuring support, would the exact requirements to the Pod Security Standards change in the future.

Also motivated by a similar issue as described in https://github.com/DataDog/helm-charts/issues/1307, although the initial scope is simply to allow for configuration of the client - any changes to Helm charts is a different matter.

### Additional Notes

Please keep in mind that `go` is not one of my primary competences, so if I've missed an obvious way to achieve something, please do correct it.

### Describe how to test/QA your changes
In a Kubernetes cluster of version `1.25` or higher do the following:
1. Install the new version `cluster-agent` configured for APM and with the following environment variable (or configuration), which follows the `Restricted` Pod Security Standard:
    * `DD_ADMISSION_CONTROLLER_AUTO_INSTRUMENTATION_INIT_SECURITY_CONTEXT="{\"capabilities\":{\"drop\":[\"ALL\"]},\"runAsNonRoot\":true,\"readOnlyRootFilesystem\":true,\"allowPrivilegeEscalation\":false,\"seccompProfile\":{\"type\":\"RuntimeDefault\"}}"` or
    * `admission_controller.auto_instrumentation.init_security_context: {"capabilities":{"drop":["ALL"]},"runAsNonRoot":true,"readOnlyRootFilesystem":true,"allowPrivilegeEscalation":false,"seccompProfile":{"type":"RuntimeDefault"}}`
3. Create a new namespace with a pod security standard `Restricted` enforced. The latter can be achieved by applying this:
    ```yaml
    apiVersion: v1
    kind: Namespace
    metadata:
      name: restricted-ns
      labels:
        pod-security.kubernetes.io/enforce: restricted
    ```
4. Next deploy an image which triggers auto instrumentation in the newly created namespace with a container, that also complies with the `Restricted` Pod Security Standard:
    ```yaml

    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: app
      namespace: restricted-ns
    spec:
      selector:
        matchLabels:
          app.kubernetes.io/name: app
      template:
        metadata:
          annotations:
            admission.datadoghq.com/<language>-lib.version: <library version>
          labels:
            app.kubernetes.io/name: app
            admission.datadoghq.com/enabled: 'true'
            tags.datadoghq.com/env: dev
            tags.datadoghq.com/service: app
            tags.datadoghq.com/version: 0.0.1
        spec:
          containers:
            - image: <image>
              name: app
              securityContext:
                allowPrivilegeEscalation: false
                capabilities:
                  drop:
                    - ALL
                readOnlyRootFilesystem: true
                runAsNonRoot: true
                seccompProfile:
                  type: RuntimeDefault
    ```
5. Verify that the pod will actually start and doesn't get stopped by the Pod Security Standard enforcement. If the same test is repeated with the existing agent the deployment will never succeed.

### CODE REVIEW

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by ASC

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 1]
```diff
@@ -774,6 +781,21 @@ func initResources() (corev1.ResourceRequirements, error) {
 	return resources, nil
 }
 
+func initSecurityContext() (*corev1.SecurityContext, error) {
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:784]: Instead of calling this method each time for each init container, which will require unmarshalling json string each time, I think it will be better to define the init container security context as a property on [the webhook struct](https://github.com/DataDog/datadog-agent/blob/5b9a9a8389e3afbc8362a8e56852f0a0a7349070/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L138-L148).

After that, you might initialise this property in the [`NewWebhook` function](https://github.com/DataDog/datadog-agent/blob/5b9a9a8389e3afbc8362a8e56852f0a0a7349070/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go#L151-L170). 
It can look like this:

```
// NewWebhook returns a new Webhook
func NewWebhook(wmeta workloadmeta.Component) (*Webhook, error) {
	filter, err := apmSSINamespaceFilter()
	if err != nil {
		return nil, err
	}

	containerRegistry := mutatecommon.ContainerRegistry("admission_controller.auto_instrumentation.container_registry")

   initSecurityContext, err := parseInitSecurityContext()
   if err != nil {
       return nil, err
   }

	return &Webhook{
		name:              webhookName,
		isEnabled:         config.Datadog().GetBool("admission_controller.auto_instrumentation.enabled"),
		endpoint:          config.Datadog().GetString("admission_controller.auto_instrumentation.endpoint"),
		resources:         []string{"pods"},
		operations:        []admiv1.OperationType{admiv1.Create},
		filter:            filter,
        initSecurityContext: initSecurityContext,
		containerRegistry: containerRegistry,
		pinnedLibraries:   getPinnedLibraries(containerRegistry),
		wmeta:             wmeta,
	}, nil
}
```

I this, I renamed your function to `parseInitSecurityContext`.

Finally, you can directly access security context as a property on the webhook object.

**Review by adel121 — COMMENTED** [review]
Hello @mschier 

Thank you for opening this PR. 🙇 

I left one comment to explain how we think the implementation can be improved for a better performance.

File: pkg/config/config_template.yaml [hunk 2]
```diff
@@ -2970,6 +2970,16 @@ api_key:
       ## Configures the memory request and limit for the init containers.
       #
       # memory:
+
+    ## @param init_security_context - json - optional
+    ## @env DD_ADMISSION_CONTROLLER_AUTO_INSTRUMENTATION_INIT_SECURITY_CONTEXT - json - optional
+    ## Security context for the init containers in json format. Follows the Kubernetes security context spec,
```
Comment (hestonhoffman) [inline pkg/config/config_template.yaml:2976]:
```suggestion
    ## Security context for the init containers in JSON format. Follows the Kubernetes security context spec,
```

File: releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    APM: In order to allow for auto instrumentation to work in Kubernetes
+    clusters that enforce a ``Restricted`` `Pod Security Standard <https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted>`_,
+    which require all containers to explicitly set a ``securityContext``,
+    an option to configure a `securityContext` to be used for all ``initContainers``
```
Comment (hestonhoffman) [inline releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml:14]:
```suggestion
    adds an option to configure a `securityContext` to be used for all ``initContainers``
```

File: releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    APM: In order to allow for auto instrumentation to work in Kubernetes
```
Comment (hestonhoffman) [inline releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml:11]:
```suggestion
    APM: In order to allow for automatic instrumentation to work in Kubernetes
```

**Review by hestonhoffman — APPROVED** [review]
Left some small documentation changes 👍

File: pkg/config/config_template.yaml (hunk 2 above)
Comment (mschier) [reply pkg/config/config_template.yaml]: Fixed

File: releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml (hunk 4 above)
Comment (mschier) [reply releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml]: Fixed

File: releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml (hunk 3 above)
Comment (mschier) [reply releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml:14]: @hestonhoffman Now that I re-read this I see that the sentence didn't make any grammatical sense, thanks for catching it 😅 
I originally intended the sentence to end with "...  have been added." instead, which I've added now - hope that's still fine? 😄

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go (hunk 1 above)
Comment (mschier) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go]: Hi @adel121 - thanks for the review! 🙏 I think your suggestion makes a lot of sense, since unmarshalling can be costly - thanks for the pointers ^^
I've implemented it, and adjusted the tests 👍

File: releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml (hunk 3 above)
Comment (hestonhoffman) [reply releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml:14]: Let's use "has been added" instead :)

Comment (mschier) [reply releasenotes/notes/apm-allow-configuration-of-security-context-for-auto-instrumentation-init-containers-28ab0a87e1920831.yaml:14]: Indeed 🙈 😅 
Fixed 😉

**Review by adel121 — COMMENTED** [review]
Left another small comment

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 5]
```diff
@@ -192,16 +193,22 @@ func NewWebhook(wmeta workloadmeta.Component) (*Webhook, error) {
 
 	containerRegistry := mutatecommon.ContainerRegistry("admission_controller.auto_instrumentation.container_registry")
 
+	initSecurityContext, err := parseInitSecurityContext()
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (mschier) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:198]: Absolutely. Added it inside `parseInitSecurityContext` to be consistent with the call to `apmSSINamespaceFilter` a few lines before this snippet - also added an extra test with another invalid json body, though still without testing the error message itself, as that doesn't seem customary in these tests?

[the remark this answers is not in the document]
Comment (adel121) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:198]: Thank you 🙇 

> though still without testing the error message itself

Yes no need to test error messages, the test you added is enough

**Comment by adel121** [comment]
LGTM to #container-platform  

Thank you @mschier for contributing to this.

**Review by adel121 — APPROVED** [review]
LGTM for #container-platform owned files. 

Thank you for @mschier for this contribution

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go [hunk 6]
```diff
@@ -199,7 +199,7 @@ func TestInjectAutoInstruConfig(t *testing.T) {
 			pod:  common.FakePodWithEnvFieldRefValue("dotnet-pod", "CORECLR_PROFILER", "path"),
 			libsToInject: []libInfo{
 				{
-					lang:  dotnet,
+					lang:  "dotnet",
```
Comment (stanistan) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go:202]: nitpick: these should be the `language` types (unnecessary change)

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go [hunk 7]
```diff
@@ -801,6 +806,65 @@ func TestInjectLibInitContainer(t *testing.T) {
...
+			lang:    java,
+			wantErr: false,
+			wantCPU: "50m",
+			wantMem: "20Mi",
+			secCtx: &corev1.SecurityContext{
+				Capabilities: &corev1.Capabilities{
+					Add:  []corev1.Capability{"NET_ADMIN", "SYS_TIME"},
+					Drop: []corev1.Capability{"ALL"},
+				},
+				Privileged: pointerOf(false),
```
Comment (stanistan) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go:824]:
```suggestion
				Privileged: pointer.Ptr(false),
```

For all of these, we already have `pointer` imported into this package, so we can drop the `pointerOf` function and use that instead

**Review by stanistan — COMMENTED** [review]
Overall this looks good, thanks for the patch, a couple of stylistic changes and I'll approve!

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go (hunk 7 above)
Comment (mschier) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go]: Ah! Perfect, I tried to find something like this, but I'm not too familiar with Go yet 😅
Thank you - I've fixed it 😄 

And thanks for the review 🙏

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go (hunk 6 above)
Comment (mschier) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go]: I've undone these now 👍 
I remember that I noticed them when merging some conflicting changes from the main branch, and these two were the only tests in the array that wasn't double quoted, see [this snippet](https://github.com/DataDog/datadog-agent/blob/bded3c7876f8ba18907c8a62f51a041b78d619fc/pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation_test.go#L185-L233), so I thought I'd "fix" it for consistency 😅
I can leave it as-is or properly fix the rest of the tests in the array if you'd like?

**Review by stanistan — APPROVED** [review]
LGTM for APM Onboarding

**Comment by mschier** [comment]
@clamoriniere @stanistan hi, I pushed a commit which I believe should fix the build and test that failed in the pipeline. There was also a failed check about binary size, which I'm not too sure about - but I don't think the binary could even build before, so that might be related to the fix as well.
I don't really know the process from here, but please let me know if I'm overstepping?

**Comment by wdhif** [comment]
> @clamoriniere @stanistan hi, I pushed a commit which I believe should fix the build and test that failed in the pipeline. There was also a [failed check about binary size](https://github.com/DataDog/datadog-agent/actions/runs/10107481177/job/27958862330), which I'm not too sure about - as far as I can tell it seems like it's caused by this PR coming from a fork? I don't really know the process from here, but please let me know if I'm overstepping?

Hey @mschier 👋 Thanks for your contribution! The issue with the binary size check could be because of the fork. Let's see if your changes fix the other checks. In the meantime, I'll checkout your work and test the binary size.

**Comment by LQss11** [comment]
Since this feature is not released yet in latest version 7.56.2 I managed to do that manually following this example:
```yaml
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    metadata:
      labels:
        app: backend
        tags.datadoghq.com/env: "restricted-example"
        tags.datadoghq.com/service: "my-service"
        tags.datadoghq.com/version: "1.0.0"
        # Must be enabled to inject env vars to containers/initContainers
        admission.datadoghq.com/enabled: "true"
      # Must be disabled since we are injecting library manually
      # annotations:
      #   admission.datadoghq.com/js-lib.version: latest        
    spec:
      ###### STEP 1: Create shared volume ######
      volumes:
        - name: datadog-auto-instrumentation
          emptyDir: {}           
      ###### STEP 2: Create datadog init container for injecting library through shared volume ######
      # Include security context manually
      initContainers:
        - name: datadog-lib-js-init
          image: gcr.io/datadoghq/dd-lib-js-init:latest
          command:
            - sh
            - copy-lib.sh
            - /datadog-lib   
          volumeMounts:
            - name: datadog-auto-instrumentation
              mountPath: /datadog-lib
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: Always   
          securityContext:
            allowPrivilegeEscalation: false
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 10000
      containers:
        - name: backend
          ###### STEP 3: get dd-trace from shared volume ######
          volumeMounts:
            - name: datadog-auto-instrumentation
              mountPath: /datadog-lib
          resources: {}
          image: myregistry.azurecr.io/nodejs-backend:latest
          env:
          ###### STEP 4: Use NODE_OPTIONS to enable trace library within the application ######
            - name: NODE_OPTIONS
              value: ' --require=/datadog-lib/node_modules/dd-trace/init'

```
